wipac_telemetry/tracing_tools/spans.py

```diff
@@ -186,81 +193,123 @@ def get_span(self, inspector: FunctionInspector) -> Span:
                         for val in func(*args, **kwargs):
                             yield val
                 except:  # noqa: E722 # pylint: disable=bare-except
                     span.end()
                     raise
-            elif conductor.behavior == SpanBehavior.END_ON_EXIT:
-                with use_span(span, end_on_exit=True):
-                    for val in func(*args, **kwargs):
-                        yield val
-            elif conductor.behavior == SpanBehavior.DONT_END:
-                with use_span(span, end_on_exit=False):
+            # CASES 2 & 3 -----------------------------------------------------
+            elif scond.behavior in (SpanBehavior.END_ON_EXIT, SpanBehavior.DONT_END):
+                end_on_exit = bool(scond.behavior == SpanBehavior.END_ON_EXIT)
+                with use_span(span, end_on_exit=end_on_exit):
                     for val in func(*args, **kwargs):
                         yield val
+            # ELSE ------------------------------------------------------------
             else:
-                raise InvalidSpanBehavior(conductor.behavior)
+                raise InvalidSpanBehavior(scond.behavior)
 
         @wraps(func)
         async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
             LOGGER.debug("Spanned Async Function")
             span = setup(args, kwargs)
+            is_iterator_class_anext_method = span.name.endswith(".__anext__")
+            reraise_stopasynciteration_outside_contextmanager = False
 
-            if conductor.behavior == SpanBehavior.ONLY_END_ON_EXCEPTION:
+            # CASE 1 ----------------------------------------------------------
+            if scond.behavior == SpanBehavior.ONLY_END_ON_EXCEPTION:
                 try:
                     with use_span(span, end_on_exit=False):
-                        return await func(*args, **kwargs)
+                        try:
+                            return await func(*args, **kwargs)
+                        except StopAsyncIteration:
+                            # intercept and temporarily suppress StopAsyncIteration
+                            if not is_iterator_class_anext_method:
+                                raise
+                            reraise_stopasynciteration_outside_contextmanager = True
                 except:  # noqa: E722 # pylint: disable=bare-except
                     span.end()
                     raise
-            elif conductor.behavior == SpanBehavior.END_ON_EXIT:
-                with use_span(span, end_on_exit=True):
-                    return await func(*args, **kwargs)
-            elif conductor.behavior == SpanBehavior.DONT_END:
-                with use_span(span, end_on_exit=False):
-                    return await func(*args, **kwargs)
+                if reraise_stopasynciteration_outside_contextmanager:
+                    raise StopAsyncIteration
+                raise RuntimeError("Malformed SpanBehavior Handling")
```

Comment:
Same thing here, I think this block (and Line 307-309) can be dropped to the end of the function.